ignite/handlers/ema_handler.py

```diff
@@ -0,0 +1,211 @@
+                  print(f"current momentum for generator: {engine.state.gen_ema_momentum}")
+                  print(f"current momentum for discriminator: {engine.state.disc_ema_momentum}")
+
+              engine.run(...)
+
+    .. versionadded:: 0.5.0
+
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        momentum: float = 0.0002,
+        momentum_warmup: Optional[float] = None,
+        warmup_iters: Optional[int] = None,
+    ) -> None:
+        if momentum_warmup is not None and not 0 < momentum_warmup < 1:
+            raise ValueError(f"Invalid momentum_warmup: {momentum_warmup}")
+        if not 0 < momentum < 1:
+            raise ValueError(f"Invalid momentum: {momentum}")
+        if momentum_warmup is not None and not momentum_warmup <= momentum:
+            raise ValueError(
+                f"momentum_warmup should be less than or equal to momentum, but got "
+                f"momentum_warmup: {momentum_warmup} and momentum: {momentum}"
+            )
+        if warmup_iters is not None and not (isinstance(warmup_iters, int) and warmup_iters > 0):
+            raise ValueError(f"Invalid warmup_iters: {warmup_iters}")
+        if not isinstance(model, nn.Module):
+            raise ValueError(
+                f"model should be an instance of nn.Module or its subclasses, but got"
+                f"model: {model.__class__.__name__}"
+            )
+        self.momentum_warmup = momentum_warmup
+        self.momentum = momentum
+        self.warmup_iters = warmup_iters
+
+        if isinstance(model, nn.parallel.DistributedDataParallel):
+            model = model.module
+        self.model = model
+
+        self.ema_model = deepcopy(self.model)
+        for param in self.ema_model.parameters():
+            param.detach_()
+        self.ema_model.eval()
+
+    def _get_momentum(self, curr_iter: int) -> float:
+        """Get current momentum, `curr_iter` should be 1-based. When `curr_iter = 1`, `momentum =
+        self.momentum_warmup`; when `curr_iter >= self.warmup_iters`, `momentum = self.momentum`"""
+
+        # TODO: use ignite's parameter scheduling, see also GitHub issue #2090
+        if curr_iter < 1:
+            raise ValueError(f"curr_iter should be at least 1, but got {curr_iter}.")
+
+        # no warmup
+        if self.momentum_warmup is None or self.warmup_iters is None:
+            return self.momentum
+
+        denominator = max(1, self.warmup_iters - 1)
+        momentum = self.momentum_warmup + (self.momentum - self.momentum_warmup) * (curr_iter - 1) / denominator
+        return min(self.momentum, momentum)
+
+    def _update_ema_model(self, engine: Engine, name: str) -> None:
+        """Update weights of ema model"""
+        momentum = getattr(engine.state, name)
+        for ema_v, model_v in zip(self.ema_model.state_dict().values(), self.model.state_dict().values()):
+            ema_v.mul_(1.0 - momentum).add_(model_v.data, alpha=momentum)
```

Comment:
I think we have to modify this part of code as the following:
```python
        for ema_v, model_v in zip(self.ema_model.state_dict().values(), self.model.state_dict().values()):
            if model_v.dtype in floating_dtypes:
                ema_v.mul_(1.0 - momentum).add_(model_v.data, alpha=momentum)
            else:
                ema_v.data.copy_(model_v.data)
```

Reply:
@vfdev-5 what about changing the terms in `zip` to `parameters()`? This should exclude the buffers. After the parameters are updated, we can then copy the buffers directly. From the test, we expect the values in buffers of EMA model and online model to be the same, right?

Reply:
Yes, actually, you are right about that. We do not need to EMA on buffers but just copy them. We can do this once while creating ema_model.

Reply:
@vfdev-5 I think we need to copy them on every update. For example, `_BatchNorm` will update their `running_mean` and `running_var` at each forward. So, if we only copy the buffers when creating the ``ema_model``. The ``ema_model`` will have default initialized buffers, e.g. `running_mean = Tensor([0, 0])`` and ``running_var = Tensor([1, 1])``.

Reply:
Actually, what we can do is to setup ema_model buffers to point to model's buffers without copying:
```
ref_model = nn.BatchNorm2d(3)
ema_model = nn.BatchNorm2d(3)

ema_model.load_state_dict(ref_model.state_dict())
for param in ema_model.parameters():
    param.detach_()
# set EMA model's BN buffers as base model BN buffers:
for m1, m2 in zip(ref_model.modules(), ema_model.modules()):
    if isinstance(m1, nn.BatchNorm2d) and isinstance(m2, nn.BatchNorm2d):
        m2.running_mean = m1.running_mean
        m2.running_var = m1.running_var

print(list(ref_model.buffers()), list(ema_model.buffers()))
> [tensor([0., 0., 0.]), tensor([1., 1., 1.]), tensor(0)] [tensor([0., 0., 0.]), tensor([1., 1., 1.]), tensor(0)]

_ = ref_model(torch.rand(12, 3, 32, 32))

print(list(ref_model.buffers()), list(ema_model.buffers()))
> [tensor([0.0502, 0.0504, 0.0499]), tensor([0.9083, 0.9084, 0.9084]), tensor(1)] [tensor([0.0502, 0.0504, 0.0499]), tensor([0.9083, 0.9084, 0.9084]), tensor(0)]

```

Reply:
It is not due to the scalar type, but due to the fact that `num_batches_tracked` is created after the buffers are copied.

In addition, the example above use `m2.running_mean` to get the buffer variable. However, if the module is wrapped in other modules. ``getattr`` seems not working anymore.
```python3
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.bn = nn.BatchNorm(2)

    def forward(self, x):
        return self.bn(x)
```
In this case, ``m = Model(); getattr(m, "bn.running_mean")`` will throw an error.

Reply:
Let me check that...